File: medium/576_out_of_boundary_paths.cpp
#include <vector>
#include <iostream>
using namespace std;

constexpr long long operator "" _b(unsigned long long l) {
    return l * 1000 * 1000 * 1000;
}

class Solution {
private:
    using value_type = unsigned long long;
    vector<vector<vector<value_type>>> memoize_table;
    int _m;
    int _n;
    int _maxMove;
public:
    int findPaths(int n, int m, int maxMove, int startRow, int startColumn) {
        _m = m;
        _n = n;
        this->memoize_table = vector<vector<vector<value_type>>>(n, vector<vector<value_type>>(m, vector<value_type>(maxMove+1, -1)));
        return f(startRow, startColumn, maxMove);
    }
private:
    value_type f(int i, int j, int k){
        if(k == 0) return 0;

        value_type key = memoize_table[i][j][k];

        if(key == -1){
            value_type paths{0};
            // Check for bounderyes
            if(i == 0) ++paths; 
            if(i == (_n - 1)) ++paths; 

            if(j == 0) ++paths; 
            if( j == (_m - 1)) ++paths; 

            if(i > 0) paths += f(i-1, j, k-1);
            if(i < _n - 1) paths += f(i+1, j, k-1);
            if(j > 0) paths += f(i, j-1, k-1);
            if(j < _m - 1) paths+= f(i, j+1, k -1);

            memoize_table[i][j][k] = paths % (1_b + 7);
            return paths;
        }
        return key;
    }
};

int main(){
    int m, n, maxMove, startRow, startColumn;
    cin >> m >> n >> maxMove >> startRow >> startColumn;
    cout << Solution().findPaths(m, n, maxMove, startRow, startColumn);
    return 0;
}

File: medium/576_out_of_boundary_paths.py
class Solution:
    def findPaths( self, n: int, m: int, maxMove: int, startRow: int, startColumn: int) -> int:
        self._n = n
        self._m = m
        self.memoize_table = [ [[-1 for _ in range(maxMove + 1)] for _ in range(m)] for _ in range(n) ]
        return self.f(startRow, startColumn, maxMove) % (10**9 + 7)

    def f(self, i: int, j: int, k: int) -> int:
        if k == 0:
            return 0
        key = self.memoize_table[i][j][k]
        if key == -1:
            paths = 0
            # Check for bounderies
            # Row boundaries
            if i == 0:
                paths += 1
            if i == self._n - 1:
                paths += 1

            # Column boundaries
            if j == 0:
                paths += 1
            if j == self._m - 1:
                paths += 1

            # Find for adjacent cells
            if i > 0:
                paths += self.f(i - 1, j, k - 1)
            if i < self._n - 1:
                paths += self.f(i + 1, j, k - 1)
            if j > 0:
                paths += self.f(i, j - 1, k - 1)
            if j < self._m - 1:
                paths += self.f(i, j + 1, k - 1)

            self.memoize_table[i][j][k] = paths
            return paths
        return key

